File: erpnext/buying/report/shipment_trip_review/shipment_trip_review.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
			//$('<div id="myModal" style="display: none; position: fixed; z-index: 9999; padding-top: 100px; left: 0; top: 0; width: 100%; height: 100%; overflow: auto; background-color: rgb(0,0,0); background-color: rgba(0,0,0,0.4);">' +
			//	'<div style="background-color: #fefefe; margin: auto; padding: 20px; border: 1px solid #888; width: 50%; height: 400px;">' +
			//		'<span id="myClose" style="color: #aaaaaa; float: right; font-size: 28px; font-weight: bold; cursor: pointer;">&times;</span>' +
			//		'<div style="background-color: #fefefe; margin: auto; padding: 20px; border: 1px solid #888; width: 100%; height: 90%;overflow-x: scroll; overflow-y: scroll;">' +
			//		'<div id="edittable"></div>' +
			//	'</div></div>' +
			//'</div>').appendTo($("body"));
			$('<div id="myModal" style="display: none; position: fixed; z-index: 9999; padding-top: 100px; left: 0; top: 0; width: 100%; height: 100%; overflow: auto; background-color: rgb(0,0,0); background-color: rgba(0,0,0,0.4);">' +
			'<div style="box-shadow: 8px 8px 10px #575757; width: 250px; margin: auto; text-align: center; background-color: #ffffff";>' +
					'<div id="edittable"></div>' +
				'</div>' +
			'</div>').appendTo($("body"));
			var modal = document.getElementById('myModal');

			// Get the button that opens the modal
			//var span = document.getElementById("myClose");
			//span.onclick = function() {
			//	modal.style.display = "none";
			//}
			window.onclick = function(event) {
				if (event.target == modal) {
					modal.style.display = "none";
				}
			}	

frappe.query_reports["Shipment Trip Review"] = {
	onload: function(report) {
		report.page.add_inner_button(__("Get Data From Invoices"), function() {
				frappe.call({
					method: "erpnext.buying.report.shipment_trip_review.shipment_trip_review.generate_data", 
					callback: (r) => {
						frappe.query_report.refresh();
					}
				});
		});
	},
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"business_unit",
			"label": __("Business Unit"),
			"fieldtype": "Link",
			"options": "Business Unit",
			"default": frappe.defaults.get_user_default("Business Unit"),
			"reqd": 1
		},
		{
			"fieldname":"exclude_completed",
			"label": __("Exclude Completed"),
			"fieldtype": "Check",
			"default": 1
		},
		{
			"fieldname":"supplier",
			"label": __("Supplier"),
			"fieldtype": "Link",
			"options": "Supplier",
		},
	],
	"formatter": function (row, cell, value, columnDef, dataContext, default_formatter) {
		//value = dataContext.account_name;
		var cell_color = "#FFFFFF"
		if (dataContext.status == 'Completed') {
			cell_color = "#A6FFA6";
			if (columnDef.df.fieldname != 'purchase_order' && columnDef.df.fieldname != 'purchase_invoice' && columnDef.df.fieldname != 'item_code') {
				columnDef.df.link_onclick = "#";
			}
			var $value = $(value).css({"display":"block","width":"100%","background-color": cell_color, "cursor": "default"});
		} else {
			if (columnDef.df.fieldname != 'purchase_order' && columnDef.df.fieldname != 'purchase_invoice' && columnDef.df.fieldname != 'item_code') {
				columnDef.df.link_onclick = "frappe.query_reports['Shipment Trip Review'].change_data('" + columnDef.df.label + "', '" + 
				columnDef.df.actualtype + "', '" + columnDef.df.actualoptions + "', '" + columnDef.df.fieldname + "', " + JSON.stringify(dataContext) + ")";
			}
		}
		if (columnDef.df.fieldname=="bank_acceptance_duration" && dataContext.bank_acceptance_duration>4) {
			cell_color = "#77FFFF";
		}
		if (columnDef.df.fieldname=="customs_release_duration" && dataContext.customs_release_duration>10) {
			cell_color = "#FFAB2D";
		}
		if (columnDef.df.fieldname=="total_duration" && dataContext.total_duration>14) {
			cell_color = "#FF9B9B";
		}
		value = default_formatter(row, cell, value, columnDef, dataContext);
		if (columnDef.df.fieldname=="qty" || columnDef.df.fieldname=="amount" || columnDef.df.fieldname=="currency") {
			var $value = $(value).css({"display":"block","width":"100%","background-color": cell_color, "cursor": "default"});
		} else {
			var $value = $(value).css({"display":"block","width":"100%","background-color": cell_color});
		}
		value = $value.wrap("<p></p>").parent().html();
		return value;
	},
	"change_data": function(fldlbl, fldtyp, opts, field, values) {
		if(values.docstatus>0) { frappe.msgprint(__("Transaction Already Submitted.")); return; }
		if( field=='qty' || field=='amount' || field=='currency') { return; }

		if (fldtyp=="Select") {
			opts = opts.split(",");
		}
		frappe.prompt({fieldname:field, options: opts, fieldtype:fldtyp,  
		label: fldlbl, reqd: 0}, function(data) {
				frappe.call({
					method: "frappe.client.set_value",
					args: {
						doctype: "Supplier Shipment Trip",
						name: values.name,
						fieldname: field,
						value: data[field],
					},
					callback: (r) => {
						frappe.query_reports['Shipment Trip Review'].add_log(field, values[field], values.name, data[field]);
					}
				});
		}, __("Change Value for Shipment  " + values.shipment_no));
	}, 
	"show_img": function(values) {
			document.getElementById("edittable").innerHTML = '<img src="' + frappe.urllib.get_base_url() + values.image + '"  style="width:100%">' +
						'<h2>'+values.employee_name+'</h2>' +
						'<p style="color: grey; font-size: 16px;">'+values.designation +'</p>' +
					'<p>' + values.employee+ '</p>' +
				'<p><button style="border: none;  outline: 0;  display: inline-block;  padding: 8px;  color: white;  background-color: #000;' + 
									'text-align: center;  cursor: pointer;  width: 100%;  font-size: 18px;" ' + 
									' onMouseOver="this.style.opacity=0.7"  onMouseOut="this.style.opacity=1"' + 
				'onclick="frappe.query_reports[\'Shipment Trip Review\'].open_form(' + values.employee + ')">View Details</button></p>';
			var modal = document.getElementById('myModal');
			modal.style.display = "block";
	}, 
	"open_form": function(doctype, docname) {
		if (!doctype || !docname) return;
		//modal.style.display = "none";

		frappe.set_route("Form", doctype, docname);
	},
	"add_log": function(field, old_value, key_name, new_value) {
		var clog0 = frappe.model.make_new_doc_and_get_name('Changes Log');
		clog = locals['Changes Log'][clog0];
		clog.report_name = frappe.query_report.report_name;
		clog.fieldname = field;
		if (old_value!="___") {
			clog.old_value = old_value;
		}
		clog.new_value = new_value;
		clog.record_key = "name";
		clog.key_name = key_name;
			frappe.call({
						method: "frappe.client.insert",
						args: {
							doc: clog
						},
						callback: function(r) {
							frappe.call({
								method: "frappe.client.submit",
								args: {
									doc: r.message
								},
								callback: function(r) {
									frappe.query_report.refresh();
								}
							});
						}
			});			
	}, 

}

File: erpnext/buying/report/shipment_trip_review/shipment_trip_review.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
	data = get_data(filters)
	columns = get_columns(filters)
	return columns, data

def get_filter_condition(filters):
	cond = ''
	if filters.get('exclude_completed') == 1:
		cond += " and ifnull(status,'')<>'Completed' "
	
	for f in filters:
		if f != 'exclude_completed':
			cond += " and " + f + " = '" + filters.get(f).replace("'", "\'") + "'"

	return cond

def get_data(filters):
	cond = get_filter_condition(filters)
	data = frappe.db.sql("""select company, business_unit, name, docstatus, ifnull(shipment_no,"___") as shipment_no, ifnull(status,'___') as status, supplier, purchase_order, purchase_invoice, 
									item_code, qty, ifnull(shipped_qty,"___") as shipped_qty, amount, currency, ifnull(destination_port,"___") as destination_port, 
									ifnull(expected_customs,"___") as expected_customs, ifnull(shipment_date,"___") as shipment_date,  
									ifnull(arrival_date,'___') as arrival_date, ifnull(docs_submition_date,'___') as docs_submition_date,    
									ifnull(acceptance_date,'___') as acceptance_date, datediff(acceptance_date,docs_submition_date) as bank_acceptance_duration,    
									ifnull(customs_start_date,'___') as customs_start_date, ifnull(customs_docs_date,'___') as customs_docs_date,    
									ifnull(samples_date,'___') as samples_date, ifnull(customs_release_date,'___') as customs_release_date, 
									datediff(customs_release_date,samples_date) as customs_release_duration, 
									ifnull(diagnoses_charges_date,'___') as diagnoses_charges_date, ifnull(final_health_release_date,'___') as final_health_release_date, 
									datediff(final_health_release_date,arrival_date) as total_duration, 
									ifnull(customs_storage_charges,'___') as customs_storage_charges, ifnull(remarks,'___') as remarks 
									from `tabSupplier Shipment Trip`
									where 1=1 %s 
									order by shipment_no,purchase_invoice, item_code""" % (cond), as_dict=True)

	return data


def get_columns(filters):
	return [
		{
			"fieldname": "company",
			"label": _("company"),
			"fieldtype": "Data",
			"hidden": 1
		},
		{
			"fieldname": "business_unit",
			"label": _("business_unit"),
			"fieldtype": "Data",
			"hidden": 1
		},
		{
			"fieldname": "name",
			"label": _("name"),
			"fieldtype": "Data",
			"hidden": 1
		},
		{
			"fieldname": "docstatus",
			"label": _("docstatus"),
			"fieldtype": "Int",
			"hidden": 1
		},
		{
			"fieldname": "shipment_no",
			"label": _("Shipment No."),
			"fieldtype": "Link",
			"actualtype": "Data",
			"options": "HHH",
			"actualoptions": "",
			"width": 120
		},
		{
			"fieldname": "status",
			"label": _("Status"),
			"fieldtype": "Link",
			"actualtype": "Select",
			"options": "HHH",
			"actualoptions": "Canceled,Completed",
			"width": 80
		},
		{
			"fieldname": "purchase_order",
			"label": _("PO"),
			"fieldtype": "Link",
			"options": "Purchase Order",
			"actualtype": "",
			"actualoptions": "",
			"width": 80
		},
		{
			"fieldname": "purchase_invoice",
			"label": _("Purchase Invoice"),
			"fieldtype": "Link",
			"options": "Purchase Invoice",
			"actualtype": "",
			"actualoptions": "",
			"width": 110
		},
		{
			"fieldname": "item_code",
			"label": _("Item"),
			"fieldtype": "Link",
			"options": "Item",
			"actualtype": "",
			"actualoptions": "",
			"width": 80
		},
		{
			"fieldname": "qty",
			"label": _("Quantity"),
			"fieldtype": "Currency",
			"width": 80
		},
		{
			"fieldname": "shipped_qty",
			"label": _("Shipped Qty"),
			"fieldtype": "Link",
			"actualtype": "Currency",
			"options": "HHH",
			"actualoptions": "",
			"width": 110
		},
		{
			"fieldname": "amount",
			"label": _("Amount"),
			"fieldtype": "Currency",
			"width": 80
		},
		{
			"fieldname": "currency",
			"label": _("Currency"),
			"fieldtype": "Link",
			"options": "Currency", 
			"width": 80
		},
		{
			"fieldname": "destination_port",
			"label": _("Destination Port"),
			"fieldtype": "Link",
			"actualtype": "Data",
			"options": "HHH",
			"actualoptions": "",
			"width": 120
		},
		{
			"fieldname": "expected_customs",
			"label": _("Expected Customs"),
			"fieldtype": "Link",
			"actualtype": "Currency",
			"options": "HHH",
			"actualoptions": "",
			"width": 120
		},
		{
			"fieldname": "shipment_date",
			"label": _("Shipment Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 110
		},
		{
			"fieldname": "arrival_date",
			"label": _("Arrival Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 80
		},
		{
			"fieldname": "docs_submition_date",
			"label": _("Bank Submition Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 150
		},
		{
			"fieldname": "acceptance_date",
			"label": _("Bank Acceptance Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 150
		},
		{
			"fieldname": "bank_acceptance_duration",
			"label": _("Bank Acceptance Duration"),
			"fieldtype": "Int",
			"width": 170
		}, 
		{
			"fieldname": "customs_start_date",
			"label": _("Customs Start Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 130
		},
		{
			"fieldname": "customs_docs_date",
			"label": _("Customs Docs Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 130
		},
		{
			"fieldname": "samples_date",
			"label": _("Samples Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 100
		},
		{
			"fieldname": "customs_release_date",
			"label": _("Customs Release Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 150
		},
		{
			"fieldname": "customs_release_duration",
			"label": _("Customs Release Duration"),
			"fieldtype": "Int",
			"width": 170
		}, 
		{
			"fieldname": "diagnoses_charges_date",
			"label": _("Diagnoses Charges Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 170
		},
		{
			"fieldname": "final_health_release_date",
			"label": _("Final Health Release Date"),
			"fieldtype": "Link",
			"actualtype": "Date",
			"options": "HHH",
			"actualoptions": "",
			"width": 170
		},
		{
			"fieldname": "total_duration",
			"label": _("Total Duration"),
			"fieldtype": "Int",
			"width": 100
		}, 
		{
			"fieldname": "customs_storage_charges",
			"label": _("Customs Storage Charges"),
			"fieldtype": "Link",
			"actualtype": "Currency",
			"options": "HHH",
			"actualoptions": "",
			"width": 170
		},
		{
			"fieldname": "remarks",
			"label": _("Remarks"),
			"fieldtype": "Link",
			"actualtype": "Text",
			"options": "HHH",
			"actualoptions": "",
			"width": 300
		},
	]

@frappe.whitelist()
def generate_data():
	frappe.db.sql("""insert ignore into `tabSupplier Shipment Trip` (`name`,  `creation`,  `modified`,  `modified_by`,  `owner`,  `docstatus`,  
					`company`,  `business_unit`, `purchase_order`,  `supplier`,  `purchase_invoice`,  `item_code`, `qty`, `shipped_qty`,  `amount`, `currency`)
					(select concat(pii.purchase_order, pii.parent,'/',pii.item_code), now(), now(), %(user)s, %(user)s, 0, 
					 pi.company, pi.business_unit, pii.purchase_order, pi.supplier, pi.name, pii.item_code,sum(pii.qty),sum(pii.qty), sum(pii.amount), pi.currency 
					 from `tabPurchase Invoice Item` pii 
					 left join `tabPurchase Invoice` pi on pi.name=pii.parent
					 left join `tabItem` i on pii.item_code=i.name
					 where i.is_stock_item=1 and pi.docstatus=1 
					 group by pii.parent,pii.item_code,pii.purchase_order )""", {"user": frappe.session.user})
